Viz cell models

cryogenian opened the pull request:
Just trying to make smaller prs for viz-cell. 

+ Renamed `Notebook.Cell.Port` to `Model.Port`
+ Mostly copied viz models from master with following changes 
  * Unified `BarConfiguration`, `LineConfiguration` and `PieConfiguration`
  * `Model.Notebook.Cell.Viz` is splitted to `Model.ChartConfiguration`, `Model.ChartAxis` and `Model.ChartSemantics`. 
  * Aesthetic changes: case -> pattern match, `s2i` -> `stringToInt`, `foo2string` -> `printFoo` etc

jonsterling commented:
Nice idea to unify those types together; I recall that there was a fair bit of duplication previously.

beckyconning commented:
I'm still reading but it seems great so far. The only thing thats bugging me is that we should probably work out a few program style standards for us to all use. This PR is similar to many parts of the package but different to many others too.

beckyconning commented on wip/src/Model/ChartSemantics.purs:
```diff
@@ -0,0 +1,293 @@
+    -- It's incorrect. Increase incorrect counter, put nothing to accum
+    | otherwise = go correct (incorrect + one) lst (Cons Nothing acc)
+
+
```
Is this extra line break on purpose to separate the instances?

cryogenian replied:
Yes. I usually make two new lines when next part of module will have some different meanings e.g. instances, or lenses. But in general I have no rules to explain why here is two newlines or one newline because I just look at the code before pr and try to make it a bit more readable. I think it useful here

beckyconning replied:
Cool I think thats a good idea.

beckyconning commented on wip/src/Model/ChartSemantics.purs:
```diff
@@ -0,0 +1,293 @@
+
+-- | This function checks values in list of `Maybe Semantics` and
+-- | returns a list with values that satisfies predicate wrapped in `Just`
+-- | or (if there more incorrect values than correct) it returns `Nothing`
```
Could you explain why the `Maybe` constructor is chosen by `correct > incorrect` rather than `correct > 0`?

cryogenian replied:
Sure, I've implemented it this way because wasn't sure about criteria of `checkPredicate` at the start. In example one can easily switch `correct > incorrect` to `toNumber correct/ toNumber incorrect > 1.3`. I know there is YAGNI, but we have no strong criteria and feature check list for this function.

garyb commented on wip/src/Model/Aggregation.purs:
```diff
@@ -0,0 +1,84 @@
+printAggregation Minimum = "∨"
+printAggregation Average = "μ"
+printAggregation Sum = "Σ"
+printAggregation Product = "Π"
```
I know this is how we currently do it, so it's not a comment on the code, just thought I'd mention it now after being reminded of this: I'm not so sure the symbolic representation of the aggregations is the most user friendly of presentations here. Statisticians will know them, but I'm not so sure the average user will know any of them aside from Σ.

Thoughts anyone?

beckyconning replied:
```purescript
printAggregation Maximum = "Maximum"
printAggregation Minimum = "Minimum"
printAggregation Average = "Average"
printAggregation Sum = "Sum"
printAggregation Product = "Product"
```

Might be easier.

beckyconning replied:
It looks cool though : ).

cryogenian replied:
I have no idea personally. We can't just rename it without remake of ui. I think that adding another two selects two current viz builder form transform it to spaceship cockpit :smile:

garyb replied:
:laughing:

beckyconning replied:
Ok cool lets leave it as is for now.

jdegoes replied:
These options exist mostly because the charting code needs some way to reduce a bunch of values into a single value when rendering the chart. We don't want users to use them, because that encourages client-side aggregation (e.g. pulling back 1 GB of data and using the "sum" aggregation to perform what would otherwise be done on the server-side with a `GROUP BY`). So in this case poor usability is a feature. :wink: 

This will go away as we evolve toward BI-lite and have visual ways to perform rollups.

beckyconning commented on wip/src/Model/ChartSemantics.purs:
```diff
@@ -0,0 +1,293 @@
+-- | This function checks values in list of `Maybe Semantics` and
+-- | returns a list with values that satisfies predicate wrapped in `Just`
+-- | or (if there more incorrect values than correct) it returns `Nothing`
+checkPredicate
```
I found this function hard to understand. I haven't compiled it yet but maybe something like this would be simpler?

```purescript
checkPredicate :: (Semantics -> Boolean) -> List (Maybe Semantics) -> Maybe (List (Maybe Semantics))
checkPredicate predicate xs = if correctCount > incorrectCount then Just filtered else Nothing
  where
  predicate :: Maybe Semantics -> Boolean
  predicate (Just semantics) | isJust $ A.elemIndex semantics nothings = false
  predicate (Just semantics) | otherwise = p semantics

  filtered :: List (Maybe Semantics)
  filtered = filter predicate xs

  correctCount :: Int
  correctCount = length filtered

  incorrectCount :: Int
  incorrectCount = (length xs) - correctCount
```

beckyconning replied:
Oh wait this isn't quite right.

beckyconning replied:
Something more like this, i'll compile it now though.

```purescript
checkPredicate :: (Semantics -> Boolean) -> List (Maybe Semantics) -> Maybe (List (Maybe Semantics))
checkPredicate predicate xs = if correctCount > incorrectCount then Just filtered else Nothing
  where
  hasValue :: Maybe Semantics -> Boolean
  hasValue (Just semantics) | isJust $ A.elemIndex semantics nothings = false

  xsWithValues :: List (Maybe Semantics)
  xsWithValues = filter hasValue xs

  filtered :: List (Maybe Semantics)
  filtered = filter predicate xsWithValues

  correctCount :: Int
  correctCount = length filtered

  incorrectCount :: Int
  incorrectCount = (length xsWithValues) - correctCount
```

beckyconning replied:
I found this function hard to understand. Maybe something broken up into simpler functions like this could do instead?

```purescript
checkPredicate :: (Semantics -> Boolean) -> List (Maybe Semantics) -> Maybe (List (Maybe Semantics))
checkPredicate predicate xs = if correctCount > incorrectCount then Just xs else Nothing
  where
  abstractValue :: Semantics -> Maybe Semantics
  abstractValue x = A.elemIndex x nothings *> pure x

  xsWithValues :: List Semantics
  xsWithValues = catMaybes $ (>>= abstractValue) <$> xs

  correct :: List Semantics
  correct = L.filter predicate xsWithValues

  correctCount :: Int
  correctCount = L.length correct

  incorrectCount :: Int
  incorrectCount = (L.length xsWithValues) - correctCount
```

beckyconning replied:
Actually now I think about it `abstractValue` should be done in a separate function before this.

beckyconning replied:
Maybe something like this?

```purescript
abstractValue :: Semantics -> Maybe Semantics
abstractValue x = A.elemIndex x nothings *> pure x

checkPredicate' :: (Semantics -> Boolean) -> List (Maybe Semantics) -> Maybe (List (Maybe Semantics))
checkPredicate' predicate xs = if correctCount > incorrectCount then Just xs else Nothing
  where
  xsWithValues :: List Semantics
  xsWithValues = catMaybes xs

  correct :: List Semantics
  correct = L.filter predicate xsWithValues

  correctCount :: Int
  correctCount = L.length correct

  incorrectCount :: Int
  incorrectCount = (L.length xsWithValues) - correctCount

-- | This function checks values in list of `Maybe Semantics` and
-- | returns a list with values that satisfies predicate wrapped in `Just`
-- | or (if there more incorrect values than correct) it returns `Nothing`
checkPredicate :: (Semantics -> Boolean) -> List (Maybe Semantics) -> Maybe (List (Maybe Semantics))
checkPredicate predicate = checkPredicate' predicate <<< map (>>= abstractValue)
```

cryogenian replied:
I have your comments in gmail, but they aren't displayed in github 0_o 
I do think that your implementation is easier to read, but it has `O(3n)` and this implementation has `O(n)` complexity and I'm afraid that this is the case where performance matters. It's possible to replace my implementation with one fold, but current tail call form is more intuitive.

beckyconning replied:
Although maybe the list should have its values and non-values abtracted before it gets passed to `checkValues` etc. I'm not sure though.

beckyconning replied:
Would you mind providing an alternative using fold? Again its because I'm inexperienced but this function whilst superbly efficient took me an hour and a half to understand.

cryogenian replied:
```purescript

type CheckAccum =
  { correct :: Int
  , incorrect :: Int
  , filtered :: List (Maybe Semantics)
  }

checkPredicateFold
  :: (Semantics -> Boolean) -> List (Maybe Semantics)
  -> Maybe (List (Maybe Semantics))
checkPredicateFold p lst =
  if result.correct > result.incorrect
  then pure $ reverse result.filtered
  else Nothing
  where
  result :: CheckAccum
  result = foldl (foldFn p) emptyAccum lst

  emptyAccum :: CheckAccum
  emptyAccum =
    { correct: zero
    , incorrect: zero
    , filtered: Nil
    }

  foldFn :: (Semantics -> Boolean) -> CheckAccum -> Maybe Semantics -> CheckAccum
  foldFn _ accum Nothing = accum { filtered = Cons Nothing accum.filtered }
  foldFn predicate accum (Just c)
    | isJust $ A.elemIndex c nothings =
        accum { filtered = Cons Nothing accum.filtered}
    | predicate c =
        accum { filtered = Cons (Just c) accum.filtered
              , correct = accum.correct + one
              }
    | otherwise =
        accum { filtered = Cons Nothing accum.filtered
              , incorrect = accum.incorrect + one
              }
```
Huh, this isn't better I suppose.

cryogenian replied:
Shall I replace current code with this?

beckyconning replied:
Yeah I prefer this :+1:

The other one looks prettier but I find this one much easier to read. Thanks!

beckyconning commented on wip/src/Model/Aggregation.purs:
```diff
@@ -0,0 +1,84 @@
+runAggregation Average nums =
+  normalize
+  $ foldl (\acc a -> bimap (add one) (add a) acc)  (Tuple zero zero) nums
+  where
```
Just wondering about people's thoughts on `where` I always thought it was meant to be indented from the thing above but this seems to work. Is this the standard? If not can we standardise on something?

cryogenian replied:
Indeed it's indented from `runAggregation`. I used to make something like this
```purescript
foo :: _ 
foo = ...
  where bar = ...
```
previously, but this format seems to be more convenient. (I've stolen it from @garyb's code)

garyb replied:
@beckyconning The `where` itself needs to be indented from its parent, but the things under a `where` can line up with the `where`.

In this case the parent is `runAggregation`, as Maxim says, not the expression starting `normalize ...` - if you think about the equivalent `let` it's perhaps clearer?

```
runAggregation Average nums =
  let normalize (Tuple count sum) = sum / count
  in normalize $ foldl (\acc a -> bimap (add one) (add a) acc)  (Tuple zero zero) nums
```

garyb replied:
(`where` is sugar for `let`, but with more pleasant indentation rules in my opinion)

beckyconning replied:
Yeah I understand that. I tend to write `where` indented from the implementation of the function but maybe thats not the best idea.

```purescript
runAggregation Average nums =
  normalize
  $ foldl (\acc a -> bimap (add one) (add a) acc)  (Tuple zero zero) nums
    where
    normalize (Tuple count sum) = sum / count
```

cryogenian replied:
Every indentation is supposed to switch context IMO. Here we switch context twice
+ to see implementation
+ to see local variables 

when `where` is indented once code says that local variables is as important as main code. I usually try to avoid nested `where` because of this -- switching between context can be very tiresome.

beckyconning replied:
I like the way you did it just wanted to make sure there was some conensus. I'll do it this way in the future too. : )

jonsterling replied:
Regarding `where` clauses, coming from Haskell practice I'm used to the things under the `where` clause being further indented than the `where` itself. I guess I don't feel strongly about it, but I would say that this is the only place where I have seen the contents of a where clause be at the same indentation level as the `where` keyboard.

Anyway, I don't actually care and am happy to do whatever folks want—but I really think that our  PR reviews can be most productive if we focus on semantic aspects rather than whitespace, so long as the code introduced is at least locally consistent.

beckyconning replied:
I agree but I do want to learn the reasons people use different styles so I can use one that works for me and the team. Maybe PRs aren't the best place for that though.

beckyconning commented on wip/src/Model/ChartAxis.purs:
```diff
@@ -0,0 +1,151 @@
+  = ValAxis (List (Maybe Semantics))
+  | CatAxis (List (Maybe Semantics))
+  | TimeAxis (List (Maybe Semantics))
+
```
One too many line breaks.

beckyconning commented on wip/src/Model/ChartAxis.purs:
```diff
@@ -0,0 +1,151 @@
+  <|> (ValAxis  <$> checkBool lst)
+  <|> (TimeAxis <$> checkTime lst)
+  <|> (CatAxis  <$> checkCategory lst)
+
```
One too many line breaks.

beckyconning commented on wip/src/Model/ChartSemantics.purs:
```diff
@@ -0,0 +1,293 @@
+            (Just <<< Value)
+            analyzeString
+
+analyzeString :: String -> Maybe Semantics
```
@jonsterling @garyb is this how you would have formatted this? would you be happy with this as the standard?

```purescript
  analyzeDate str
    <|> analyzeNumber str
    <|> analyzeMoney str
    <|> analyzePercent str
    <|> (Just $ Category str)
```

or

```purescript
  analyzeDate str <|> analyzeNumber str <|> analyzeMoney str <|> analyzePercent str <|> (Just $ Category str)
```

Is how I would have written it but I'm not very experienced and the way @cryogenian has written this here has a bunch of benefits.

garyb replied:
Previously I would have written it as 

``` purescript
analyzeString str = analyzeDate str
                <|> analyzeNumber str
                <|> analyzeMoney str
                <|> analyzePercent str
                <|> (Just $ Category str)
```

But that's not okay according to the rules of Jondentation :wink:. I think Maxim's current formatting or any of the suggestions are all fine, although I slightly prefer either single line or Maxim's style, as it doesn't emphasise the `analyzeDate` case quite so much. @jonsterling may have the definitive answer!

beckyconning commented on wip/src/Model/ChartAxis.purs:
```diff
@@ -0,0 +1,151 @@
+
+
+  where
+  transpose :: Json -> JArray
```
I still can't really read these, maybe it will come with experience but willingness to do bounties is a potential positive of writing these in a way less experienced developers can read.

I would have thought this could be broken into multiple referenced functions but I could be wrong.

beckyconning replied:
e.g. what does the function `tpl2sm >>> Sm.fromList >>> encodeJson` do? 

could it be given a name?

beckyconning replied:
Fill in the blanks : )

```
transpose :: Json -> JArray
transpose = decodeJson >>> map (Sm.toList >>> map f) >>> g
  where
  --g :: 
  --g = 

  --f ::
  --f ::
```

garyb replied:
To me, just reading it without examining the context I see "a chain of operations. start with a tuple and convert to `StrMap`, create `StrMap` from list, encode the list as JSON"

Having read what's going on, that isn't quite right as `tpl2sm` doesn't produce a `StrMap`, it produces a list of tuples that are compatible with `SM.fromList`, but anyway I'm not so sure this piece can be given a particularly meaningful name, as the whole thing here is the transpose operation - it's inherently complicated, especially when the data is not encoded in a "natural" form (meaning, we're having to encode/decode json while we're at it).

cryogenian replied:
I think `tpl2sm >>> Sm.fromList >>> encodeJson` has no much sense in fact, it just function that works with unnamed list of tuples and emits `JObject`. 
I mean this is a bit different way of modeling function as far as I can tell one can use flow of data or steps with meaningful variables.

beckyconning replied:
What does

```purescript
Sm.toList >>> map (tpl2sm >>> Sm.fromList >>> encodeJson)
```

actually do? It seems to convert one big string map into many json encoded string maps. Why do we want that rather than the initial string map?

beckyconning replied:
also maybe `fromList` should be inside `tupleToStringMap`

cryogenian replied:
User requested this feature. It used to provide the ability to make charts for this kind of data `{"foo": 1, "bar": 2}`

beckyconning replied:
How can I get a `JArray` to test this function with?

cryogenian replied:
`[ encodeJson $ Sm.fromList $ L.toList [ Tuple "foo" 1 ]]`

beckyconning commented on wip/src/Model/ChartAxis.purs:
```diff
@@ -0,0 +1,151 @@
+
+
+analyzeJArray :: JArray -> M.Map JCursor Axis
+analyzeJArray arr =
```
Again I can't really read this. Can it be split into simpler functions and recomposed?

beckyconning commented on wip/src/Model/ChartConfiguration.purs:
```diff
@@ -0,0 +1,70 @@
+_measures = _ChartConfiguration <<< lens _.measures _{measures = _}
+
+instance encodeJsonChartConfiguration :: EncodeJson ChartConfiguration where
+  encodeJson (ChartConfiguration r) = "series" := r.series
```
I would prefer a line break after `=` to avoid this empty indentation.

cryogenian replied:
I'm not sure why but it doesn't compile with such indentation.

beckyconning replied:
Oh weird, it must be an instance thing. Thanks for trying anyway.

cryogenian replied:
I was tried it initially when I start to code this part, will check if it works now.

cryogenian replied:
Yep, it works, will change.

cryogenian commented:
+ Removed unneeded newlines from `ChartAxis`
+ Changed from recursion to fold in `checkPredicate`
+ Changed indentation of `EncodeJson` instances. 
+ Haven't changed `analyzedJArray`. I think the only way to make temp variables named is completely rewriting this function, but it works and I think it's not so bad. 

@beckyconning is it good enough to merge?

beckyconning commented on wip/src/Model/ChartAxis.purs:
```diff
@@ -0,0 +1,142 @@
+                      >>> encodeJson))
+    -- if decoding to StrMap has no success return empty JArray it will
+    -- not produce any Axises. Else convert List of JObjects to Array
+    >>> either (const []) fromList
```
so i'm trying to test this function to get a better idea of what it does

```purescript
  transpose :: Json -> JArray
  transpose = decodeJson >>> map g >>> either (const []) fromList

  f = Sm.toList >>> map (tpl2sm >>> Sm.fromList >>> encodeJson)

  g x = traceAny x (\_ -> traceAny (f x) (\_ -> f x))
```

but i can't think of how to produce some input for it. any ideas?

cryogenian replied:
It takes something like this 
`[encodeJson $ Sm.fromList $ L.toList [ Tuple "foo" 1, Tuple "bar" 2, Tuple "baz" 3]]` and returns 
```
[ encodeJson $ Sm.fromList $ L.toList [ Tuple "key" "foo", Tuple "value" 1]
, encodeJson $ Sm.fromList $ L.toList [ Tuple "key" "bar", Tuple "value" 2]
, encodeJson $ Sm.fromList $ L.toList [ Tuple "key" "baz", Tuple "value" 3] 
]
```

beckyconning replied:
Could I suggest

```purescript
  transpose :: Json -> JArray
  transpose =
    decodeJson
      >>> map splitStringMapIntoEncodedStringMaps
      >>> either (const []) fromList

  splitStringMapIntoEncodedStringMaps :: _
  splitStringMapIntoEncodedStringMaps = Sm.toList >>> map (tpl2sm >>> encodeJson)

  tupleToStringMap :: Tuple String Json -> _
  tupleToStringMap (Tuple key val) =
    Sm.fromList
      $ Cons (Tuple "key" $ encodeJson key)
      $ Cons (Tuple "value" val)
      $ Nil
```

cryogenian replied:
```purescript
toKeyValJArray :: Json -> JArray
toKeyValJArray =
  decodeJson >>> (map toKeyValJsons) >>> either (const []) fromList

toKeyValJsons :: JObject -> List Json
toKeyValJsons =
  Sm.toList >>> map (toKeyVals >>> Sm.fromList >>> encodeJson)

toKeyVals :: Tuple String Json -> List (Tuple String Json)
toKeyVals =
  Cons (Tuple "key" $ encodeJson key)
  $ Cons (Tuple "value" val) Nil
```
How does this look?

beckyconning replied:
Perfect! Thank you! : )

jonsterling commented:
Regarding indentation and stuff, my position is we shouldn't worry too much so long as a module is internally consistent (within reason).

cryogenian commented:
+ Added names for local functions in `analyzeJArray`

cryogenian commented:
I strongly agree with @jonsterling as far as module is consistent and code works we shouldn't worry much about naming, indentation and other things.
